### PR TITLE
CSI-5181 add new vgc controller - utils files

### PR DESCRIPTION
**Pull request by ArbelNathan**
related PRs - #35

### CODE REVIEW

File: controllers/utils/utils.go [hunk 1]
```diff
@@ -17,7 +17,10 @@ limitations under the License.
 package utils
 
 import (
+	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"reflect"
+	"time"
```
Comment (matancarmeli7) [inline controllers/utils/utils.go:23]: can you organize the imports to the the core ones will be above and after new line the pulled ones will be after it, like [here](https://blog.devgenius.io/sort-go-imports-acb76224dfa7)

File: controllers/utils/volumegroup.go [hunk 2]
```diff
@@ -19,17 +19,47 @@ package utils
...
+	err := client.Get(context.TODO(), namespacedVGC, vgc)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Error(err, "VolumeGroupContent not found", "VolumeGroupContent Name", vgName)
+		}
+		return nil, err
+	}
+
+	return vgc, nil
+}
```
Comment (matancarmeli7) [inline controllers/utils/volumegroup.go:48]:
```suggestion
func GetVG(client client.Client, logger logr.Logger, vgName string, vgNamespace string) (*volumegroupv1.VolumeGroup, error) {
	logger.Info(fmt.Sprintf(messages.GetVG, vgName, vgNamespace))
	vg := &volumegroupv1.VolumeGroup{}
	namespacedVG := types.NamespacedName{Name: vgName, Namespace: vgNamespace}
	err := client.Get(context.TODO(), namespacedVG, vg)
	if err != nil {
		if errors.IsNotFound(err) {
			logger.Error(err, "VolumeGroup not found", "VolumeGroup Name", vgName)
		}
		return nil, err
	}

	return vg, nil
}
```
and we don't have `messages.GetVG` so we need to create one

File: controllers/utils/volumegroup.go [hunk 3]
```diff
@@ -202,7 +232,7 @@ func removeFromPVCList(pvc *corev1.PersistentVolumeClaim, pvcList []corev1.Persi
 }
 
 func getVgId(logger logr.Logger, client client.Client, vg *volumegroupv1.VolumeGroup) (string, error) {
-	vgc, err := GetVGC(client, logger, GetStringField(vg.Spec.Source, "VolumeGroupContentName"), vg.Name, vg.Namespace)
+	vgc, err := GetVGC(client, logger, GetStringField(vg.Spec.Source, "VolumeGroupContentName"), vg.Namespace)
```
Comment (matancarmeli7) [inline controllers/utils/volumegroup.go:235]: why did you remove the `vg.Name`? the log can be look strange

File: controllers/utils/volumegroupcontent.go [hunk 4]
```diff
@@ -19,11 +19,11 @@ package utils
 import (
 	"context"
 	"fmt"
+	"github.com/IBM/csi-volume-group-operator/controllers/volumegroup"
+	csi "github.com/IBM/csi-volume-group/lib/go/volumegroup"
```
Comment (matancarmeli7) [inline controllers/utils/volumegroupcontent.go:23]: also here about the imports

File: controllers/utils/volumegroupcontent.go [hunk 5]
```diff
@@ -233,24 +252,65 @@ func vgcRetryOnConflictFunc(client client.Client, vgc *volumegroupv1.VolumeGroup
...
+	}
+	updateThinVGCSpec(vgc)
+	if err = UpdateObject(client, vgc); err != nil {
+		return err
+	}
+	return nil
+}
+
+func updateThinVGCSpec(vgc *volumegroupv1.VolumeGroupContent) {
+	if GetStringField(vgc.Spec, "VolumeGroupDeletionPolicy") == "" {
```
Comment (matancarmeli7) [inline controllers/utils/volumegroupcontent.go:302]: I think if you would change it to: `if vgc.Spec.VolumeGroupDeletionPolicy { ...` it will work also, can you do I quick check for it?

File: pkg/messages/controller_messages.go [hunk 6]
```diff
@@ -58,4 +60,5 @@ var (
 	RetryUpdateVGStatus            = "Retry update %s/%s volumeGroup status due to conflict error"
 	RetryUpdateVGCtStatus          = "Retry update %s/%s volumeGroupContent status due to conflict error"
 	RetryUpdateFinalizer           = "Retry update finalizer due to conflict error"
+	VgIsStillExist                 = "cant delete %s/%s volumeGroupContent because volumeGroup is still exist"
```
Comment (matancarmeli7) [inline pkg/messages/controller_messages.go:63]:
```suggestion
	VgIsStillExist                 = "Cant delete %s/%s volumeGroupContent because volumeGroup is still exist"
```

File: controllers/utils/volumegroup.go (hunk 2 above)
Comment (ArbelNathan) [reply controllers/utils/volumegroup.go]: done

File: pkg/messages/controller_messages.go (hunk 6 above)
Comment (ArbelNathan) [reply pkg/messages/controller_messages.go]: done

File: controllers/utils/volumegroup.go (hunk 3 above)
Comment (matancarmeli7) [reply controllers/utils/volumegroup.go:234]: ok

File: controllers/utils/modify_volumegroup.go [hunk 7]
```diff
@@ -18,7 +18,6 @@ package utils
 
 import (
 	"fmt"
-
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controllers/utils/modify_volumegroup.go]: done

File: controllers/utils/utils.go (hunk 1 above)
Comment (ArbelNathan) [reply controllers/utils/utils.go]: done

File: controllers/utils/volumegroupcontent.go (hunk 4 above)
Comment (ArbelNathan) [reply controllers/utils/volumegroupcontent.go]: done

File: controllers/utils/volumegroupcontent.go (hunk 5 above)
Comment (ArbelNathan) [reply controllers/utils/volumegroupcontent.go:303]: I checked, its not working.

**Review by matancarmeli7 — APPROVED** [review]
LGTM

File: controllers/utils/handle_messages.go [hunk 8]
```diff
@@ -70,6 +70,7 @@ func HandleVGCErrorMessage(logger logr.Logger, client client.Client, vgc *volume
 		if uErr := createNamespacedObjectErrorEvent(logger, client, vgc, errorMessage, reason); uErr != nil {
 			return uErr
 		}
+		return err
```
Comment (erantzabari) [inline controllers/utils/handle_messages.go:73]: Instead of return err inside the if - return err at the end of the function (instead of nil)
That apply to all functions in this file

File: controllers/utils/volumegroupcontent.go [hunk 9]
```diff
@@ -221,6 +228,19 @@ func updateVGCStatusPVList(client client.Client, vgc *volumegroupv1.VolumeGroupC
 	return nil
 }
 
+func UpdateVGCStatusError(client client.Client, vgc *volumegroupv1.VolumeGroupContent, logger logr.Logger, message string) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		vgc.Status.Error = &volumegroupv1.VolumeGroupError{Message: &message}
+		err := vgcRetryOnConflictFunc(client, vgc, logger)
+		return err
+	})
+	if err != nil {
```
Comment (erantzabari) [inline controllers/utils/volumegroupcontent.go:237]: You can just return err, no need of the if

File: controllers/utils/volumegroupcontent.go [hunk 10]
```diff
@@ -233,24 +253,65 @@ func vgcRetryOnConflictFunc(client client.Client, vgc *volumegroupv1.VolumeGroup
...
+	vgc.Spec.VolumeGroupDeletionPolicy = getVolumeGroupDeletionPolicy(vgClass)
+}
+
+func UpdateThinVGC(client client.Client, vgcNamespace, vgcName string, logger logr.Logger) error {
+	vgc, err := GetVGC(client, logger, vgcName, vgcNamespace)
+	if err != nil {
+		return err
+	}
+	updateThinVGCSpec(vgc)
+	if err = UpdateObject(client, vgc); err != nil {
```
Comment (erantzabari) [inline controllers/utils/volumegroupcontent.go:296]: Return err - no need to check for nil

**Review by erantzabari — APPROVED** [review]

File: controllers/utils/handle_messages.go (hunk 8 above)
Comment (ArbelNathan) [reply controllers/utils/handle_messages.go]: done

File: controllers/utils/volumegroupcontent.go (hunk 9 above)
Comment (ArbelNathan) [reply controllers/utils/volumegroupcontent.go]: done

File: controllers/utils/volumegroupcontent.go (hunk 10 above)
Comment (ArbelNathan) [reply controllers/utils/volumegroupcontent.go:292]: I prefer to leave like this.
the error handle here is a break. if we remove it and add stage after UpdateObject we could forget to break in case of an error from UpdateObject.

**Review by matancarmeli7 — DISMISSED** [review]

**Review by matancarmeli7 — APPROVED** [review]

**Review by matancarmeli7 — APPROVED** [review]
